### PR TITLE
feat: Implement missing stuff for scheduled events

### PR DESCRIPTION
**Pull request by Damego**
## Pull Request Type

- [x] Feature addition
- [ ] Bugfix
- [ ] Documentation update
- [ ] Code refactor
- [ ] Tests improvement
- [ ] CI/CD pipeline enhancement
- [ ] Other: [Replace with a description]

## Description
This library is missing support for 5 events of scheduled events

## Changes
- Created models for [each event](https://discord.com/developers/docs/topics/gateway-events#guild-scheduled-event-create).
- Created processor for scheduled events
- Cache scheduled events
- Added helper methods to cache scheduled events
- Refactored code to use cache methods
- Added `Client.get_scheduled_event`

## Test Scenarios
Create/update/delete scheduled event + add/remove user from it via discord client or this library.
Make sure you have enabled `GUILD_SCHEDULED_EVENTS` intent.

Create listener for each event
```py
@interactions.listen()
async def create(event: events.GuildScheduledEventCreate):
    print(event.scheduled_event)


@interactions.listen()
async def update(event: events.GuildScheduledEventUpdate):
    print(event.before)
    print(event.after)


@interactions.listen()
async def delete(event: events.GuildScheduledEventDelete):
    print(event.scheduled_event)


@interactions.listen()
async def user_add(event: events.GuildScheduledEventUserAdd):
    print(event.scheduled_event)
    print(event.user)


@interactions.listen()
async def user_remove(event: events.GuildScheduledEventUserRemove):
    print(event.scheduled_event)
    print(event.user)
```

## Python Compatibility

- [ ] I've ensured my code works on Python `3.10.x`
- [ ] I've ensured my code works on Python `3.11.x`

## Checklist

- [x] I've run the `pre-commit` code linter over all edited files
- [ ] I've tested my changes on supported Python versions
- [ ] I've added tests for my code, if applicable
- [x] I've updated / added  documentation, where applicable

### CODE REVIEW

File: interactions/api/events/discord.py [hunk 1]
```diff
@@ -751,3 +756,46 @@ class GuildAuditLogEntryCreate(GuildEvent):
...
+
+
+@attrs.define(eq=False, order=False, hash=False, kw_only=False)
+class GuildScheduledEventUserAdd(GuildEvent):
+    """Dispatched when scheduled event is created"""
+
+    scheduled_event: Optional[interactions.models.ScheduledEvent]
+    """The scheduled event object if cached"""
+    user: "User"
+    """The user that has been added/removed from scheduled event"""
```
Comment (AstreaTSS) [inline interactions/api/events/discord.py]: Why is this not marked optional? This gets the user from its cache, which has a decent chance of not having anything and so returning `None`. I don't believe interactions.py caches any user object right before either.

Comment (Damego) [reply interactions/api/events/discord.py]: Yeah you're right. I just missed it. Will resolve tomorrow

File: interactions/api/events/discord.py [hunk 2]
```diff
@@ -751,3 +756,56 @@ class GuildAuditLogEntryCreate(GuildEvent):
 
     audit_log_entry: interactions.models.AuditLogEntry = attrs.field(repr=False)
     """The audit log entry object"""
+
+
+@attrs.define(eq=False, order=False, hash=False, kw_only=False)
+class GuildScheduledEventCreate(BaseEvent):
+    """Dispatched when scheduled event is created"""
+
+    scheduled_event: interactions.models.ScheduledEvent = attrs.field(repr=True)
```
Comment (i0bs) [inline interactions/api/events/discord.py]: We're using the `Absent["type"]`/`Optional["type"]`/`"type"` convention for string-type parsing everywhere else in our code. Sequential proposals suggest moving over to it

```suggestion
    scheduled_event: "ScheduledEvent" = attrs.field(repr=True)
```

File: interactions/api/events/discord.py [hunk 3]
```diff
@@ -751,3 +756,56 @@ class GuildAuditLogEntryCreate(GuildEvent):
...
+    """The scheduled event object"""
+
+
+@attrs.define(eq=False, order=False, hash=False, kw_only=False)
+class GuildScheduledEventUpdate(BaseEvent):
+    """Dispatched when scheduled event is updated"""
+
+    before: Absent[interactions.models.ScheduledEvent] = attrs.field(repr=True)
```
Comment (i0bs) [inline interactions/api/events/discord.py]:
```suggestion
    before: Absent["ScheduledEvent"] = attrs.field(repr=True)
```

File: interactions/api/events/discord.py [hunk 4]
```diff
@@ -751,3 +756,56 @@ class GuildAuditLogEntryCreate(GuildEvent):
...
+    """The scheduled event before this event was created"""
+    after: interactions.models.ScheduledEvent = attrs.field(repr=True)
```
Comment (i0bs) [inline interactions/api/events/discord.py]:
```suggestion
    after: "ScheduledEvent" = attrs.field(repr=True)
```

File: interactions/api/events/discord.py [hunk 5]
```diff
@@ -751,3 +756,56 @@ class GuildAuditLogEntryCreate(GuildEvent):
...
+class GuildScheduledEventUserAdd(GuildEvent):
+    """Dispatched when scheduled event is created"""
+
+    scheduled_event_id: "Snowflake_Type" = attrs.field(repr=True)
+    """The ID of the scheduled event"""
+    user_id: "Snowflake_Type" = attrs.field(repr=True)
+    """The ID of the user that has been added/removed from scheduled event"""
+
+    @property
+    def scheduled_event(self) -> Optional[interactions.models.ScheduledEvent]:
```
Comment (i0bs) [inline interactions/api/events/discord.py]:
```suggestion
    def scheduled_event(self) -> Optional["ScheduledEvent"]:
```

File: interactions/client/client.py [hunk 6]
```diff
@@ -2282,10 +2289,29 @@ async def fetch_scheduled_event(
...
             scheduled_event_data = await self.http.get_scheduled_event(guild_id, scheduled_event_id, with_user_count)
-            return ScheduledEvent.from_dict(scheduled_event_data, self)
+            return self.cache.place_scheduled_event_data(scheduled_event_data)
         except NotFound:
             return None
 
+    def get_scheduled_event(
+        self,
+        scheduled_event_id: "Snowflake_Type",
+    ) -> Optional["ScheduledEvent"]:
```
Comment (i0bs) [inline interactions/client/client.py:2299]: (As seen, we're doing it here)

**Review by i0bs — CHANGES_REQUESTED** [review]
Please check to ensure imports normally work for `"ScheduledEvent"` type signature

**Review by i0bs — APPROVED** [review]

**Review by silasary — APPROVED** [review]

**Pull request merged**